File: rust/src/extract/outline.rs
//! Hierarchical outline extractor. Mirrors src/skimr/extract/outline.py.
//!
//! Detects sections via structural heading patterns (markdown, allcaps,
//! short-colon-label). For each section picks the highest-composite-score
//! non-heading sentence as the representative.
//!
//! Note: `outline()` uses a narrower heading predicate than
//! `crate::headings::is_heading` — it does NOT treat "fewer than 4 content
//! tokens" as a heading signal, because that heuristic misclassifies short
//! body sentences as headings, leaving sections with no
//! representative-sentence candidates.

// Heading patterns + helpers live in `crate::headings` — single source of
// truth so a new pattern requires one edit, not two. T13b Class D cases
// intentionally NOT handled (still tracked as follow-ups):
//   - "Meeting: Platform Migration Planning" (Label:Subject inline pattern)
//   - "Held: Section 412(b) does not authorize..." (inline colon-label
//     with body text on the same line)
//   - "Reply from support (Kai T., day 1)" (parenthetical structured
//     heading; too corpus-specific for a general predicate)
use crate::headings::{heading_name, is_structural_heading, md_depth};
use crate::sentences::split_sentences;
use crate::tfidf::{composite_score_parts, separate_heading_lines};

#[derive(Debug, Clone, PartialEq)]
pub struct Section {
    pub depth: usize,
    pub name: String,
    pub representative_sentence: String,
}

struct Sect {
    depth: usize,
    name: String,
    body: Vec<usize>,
}

#[must_use]
pub fn outline(text: &str) -> Vec<Section> {
    if text.is_empty() {
        return Vec::new();
    }

    // Pre-split heading-only lines so they become standalone sentences —
    // mirrors tfidf::summarize's default-mode preprocessing.
    let prepared = separate_heading_lines(text);
    let sentences = split_sentences(&prepared);
    if sentences.is_empty() {
        return Vec::new();
    }

    let parts = composite_score_parts(&sentences);
    let scores: Vec<f64> = parts
        .iter()
        .map(|(t, p, l)| 0.60 * t + 0.25 * p + 0.15 * l)
        .collect();

    let mut sections: Vec<Sect> = Vec::new();
    let mut current: Option<Sect> = None;
    for (i, s) in sentences.iter().enumerate() {
        if is_structural_heading(s) {
            if let Some(cur) = current.take() {
                sections.push(cur);
            }
            let name = heading_name(s).unwrap_or_default();
            current = Some(Sect {
                depth: md_depth(s),
                name,
                body: Vec::new(),
            });
        } else if let Some(cur) = current.as_mut() {
            cur.body.push(i);
        }
    }
    if let Some(cur) = current {
        sections.push(cur);
    }

    let mut out = Vec::with_capacity(sections.len());
    for sect in sections {
        if sect.body.is_empty() || sect.name.is_empty() {
            continue;
        }
        let best_idx = *sect
            .body
            .iter()
            .max_by(|a, b| {
                // NaN-tolerant: degrade to Equal rather than panicking on a
                // corrupt score. The .then_with tiebreak still runs and
                // produces a deterministic order.
                scores[**a]
                    .partial_cmp(&scores[**b])
                    .unwrap_or(std::cmp::Ordering::Equal)
                    .then_with(|| b.cmp(a))
            })
            .expect("non-empty");
        out.push(Section {
            depth: sect.depth,
            name: sect.name,
            representative_sentence: sentences[best_idx].clone(),
        });
    }
    out
}

/// Lightweight table of contents — section names in document order.
///
/// Equivalent to `outline(text).into_iter().map(|s| s.name).collect()` but
/// exposed as its own primitive for discoverability. Uses the regex heading
/// detector; no backend parameter (regex is the only option, same as
/// `outline()`).
#[must_use]
pub fn toc(text: &str) -> Vec<String> {
    outline(text).into_iter().map(|s| s.name).collect()
}

File: src/lede/extract/outline.py
"""Hierarchical outline extractor.

Detects sections via structural heading patterns (markdown, allcaps,
short-colon-label). For each section picks the highest-composite-score
non-heading sentence as the representative.

Note: outline() uses a narrower heading predicate than _headings.is_heading
— it does NOT treat "fewer than 4 content tokens" as a heading signal,
because that heuristic misclassifies short body sentences as headings,
leaving sections with no representative-sentence candidates. Outline
needs real section structure, not title-length prose.

Mirrors rust/src/extract/outline.rs.
"""
import re

from .._headings import heading_name, is_structural_heading, md_depth
from .._types import Section
from ..sentences import split_sentences
from ..tfidf import _composite_score_parts, _separate_heading_lines

# Heading patterns + helpers live in lede._headings — single source of
# truth so a new pattern requires one edit, not two. T13b Class D cases
# intentionally NOT handled (still tracked as follow-ups):
#   - "Meeting: Platform Migration Planning" (Label:Subject inline pattern)
#   - "Held: Section 412(b) does not authorize..." (inline colon-label
#     with body text on the same line)
#   - "Reply from support (Kai T., day 1)" (parenthetical structured
#     heading; too corpus-specific for a general predicate)

# Local aliases so existing call sites keep their names; both come from
# the shared module.
_is_structural_heading = is_structural_heading
_md_depth = md_depth


def outline(text: str) -> tuple[Section, ...]:
    """Extract sections: heading + top-scoring non-heading sentence within."""
    if not text:
        return ()

    # Pre-split heading-only lines so they become standalone sentences —
    # mirrors tfidf.summarize's default-mode preprocessing.
    prepared = _separate_heading_lines(text)
    sentences = split_sentences(prepared)
    if not sentences:
        return ()

    parts = _composite_score_parts(sentences)
    scores = [0.60 * t + 0.25 * p + 0.15 * l for (t, p, l) in parts]

    # Group non-heading sentences under the most recent heading.
    sections: list[tuple[int, str, list[int]]] = []
    current: tuple[int, str, list[int]] | None = None
    for i, s in enumerate(sentences):
        if _is_structural_heading(s):
            if current is not None:
                sections.append(current)
            name = heading_name(s) or ""
            current = (_md_depth(s), name, [])
        else:
            if current is not None:
                current[2].append(i)
    if current is not None:
        sections.append(current)

    out: list[Section] = []
    for depth, name, body in sections:
        if not body or not name:
            continue
        best_idx = max(body, key=lambda i: scores[i])
        out.append(Section(
            depth=depth,
            name=name,
            representative_sentence=sentences[best_idx],
        ))
    return tuple(out)


def toc(text: str) -> tuple[str, ...]:
    """Return section names in document order. Lightweight TOC.

    Equivalent to ``tuple(s.name for s in outline(text))`` but exposed as
    its own primitive for discoverability. Uses the regex heading detector;
    no backend kwarg because the only backend is regex (parity with
    ``outline()``).
    """
    return tuple(s.name for s in outline(text))


# Register 'regex' backend entry for registry symmetry with metadata/phrases/
# correlate_facts. The public `toc()` function does not accept a backend kwarg
# (regex is the only option, same as outline); registration keeps the registry
# complete so any introspection shows toc.
from ._backends import register as _register  # noqa: E402

_register("regex", "toc", toc)
